src/libutils/oglapphelpers/metalapp.mm

```diff
@@ -0,0 +1,391 @@
+
+MetalApp::MetalApp(const char * winTitle, int winWidth, int winHeight)
+    : ScreenApp(winTitle, winWidth, winHeight)
+{
+    initContext();
+}
+
+void MetalApp::initContext()
+{
+    NSOpenGLContext* currentContext = [NSOpenGLContext currentContext];
+    
+    m_context = std::unique_ptr<GraphicsContext>(new GraphicsContext(currentContext, MTLCreateSystemDefaultDevice()));
+}
+
+MetalApp::~MetalApp()
+{
+    m_metalBuilder.reset();
+}
+
+void MetalApp::initImage(int imgWidth, int imgHeight, Components comp, const float * image)
+{
+    std::vector<float> imageData;
+    if(comp == Components::COMPONENTS_RGB)
+    {
+        imageData = RGB_to_RGBA(image, 3 * imgWidth * imgHeight);
+        image = imageData.data();
+    }
+    
+    setImageDimensions(imgWidth, imgHeight, comp);
+    m_image = std::make_shared<MtlTexture>(m_context->metalDevice, imgWidth, imgHeight, image);
+    m_outputImage = std::make_shared<MtlTexture>(m_context->metalDevice, m_context->glContext, imgWidth, imgHeight, image);
+}
+
+void MetalApp::updateImage(const float *image)
+{
+    std::vector<float> imageData;
+    if(getImageComponents() == Components::COMPONENTS_RGB)
+    {
+        imageData = RGB_to_RGBA(image, 3 * m_image->getWidth() * m_image->getHeight());
+        image = imageData.data();
+    }
+    m_image->update(image);
+}
+
+void MetalApp::prepareAndBindOpenGLState()
+{
+    // A dummyShaderDesc is enough.
+    // The builder will only be used to build GL program
+    GpuShaderDescRcPtr dummyShaderDesc = GpuShaderDesc::CreateShaderDesc();
+    m_oglBuilder = OpenGLBuilder::Create(dummyShaderDesc);
+    
+    std::ostringstream main;
+        
+    main <<    std::endl
+            << "uniform sampler2DRect img;" << std::endl
+            << std::endl
+            << "void main()" << std::endl
+            << "{" << std::endl
+            << "    gl_FragColor = texture2DRect(img, gl_TexCoord[0].st * "
+            << "vec2(" << m_outputImage->getWidth() << ", " << m_outputImage->getHeight() << "));"
+            << std::endl
+            << "}" << std::endl;
+    
+    
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_RECTANGLE_EXT, m_outputImage->getGLHandle());
+    
+    // Build the fragment shader program.
+    m_oglBuilder->buildProgram(main.str().c_str(), true);
+
+    // Enable the fragment shader program, and all needed resources.
+    m_oglBuilder->useProgram();
+    
+    // The image texture.
+    glUniform1i(glGetUniformLocation(m_oglBuilder->getProgramHandle(), "img"), 0);
+    
+    m_glStateBound = true;
+}
+
+void MetalApp::setShader(GpuShaderDescRcPtr & shaderDesc)
+{
+    std::ostringstream main;
+    
+    m_metalBuilder = MetalBuilder::Create(shaderDesc);
+    m_metalBuilder->allocateAllTextures(1);
+    
+    if(shaderDesc->getLanguage() == GPU_LANGUAGE_MSL_2_0)
+    {
+        std::ostringstream uniformParams;
+        std::ostringstream params;
+        
+        main << "//Metal Shading Language version 2.3\n"
```

Comment:
The MSL translation is based on MSL 2.0 but the comment indicates MSL 2.3. The generated shader program should also be 2.0 to be consistent. Is there a need to be 2.3?

Reply:
Wow!!! Good Catch :D